### DISCUSSION TITLE
Pydantic + Dataclass + Generic = ??

### DISCUSSION THREAD

**Discussion by TheRealQuantam** [pydantic/pydantic#5046] [Question]
Is it possible to have a generic model or dataclass that takes a list of parameters rather than a dict? GenericModel provides model + generic, but does not work with positional arguments. dataclass accepts positional arguments, but I haven't been able to figure out how to get it to work with generics. NamedTuple works with positional arguments and generics, but I haven't been able to figure out how to attach validators to it.

**Comment by samuelcolvin** [comment]
Do standard lib dataclasses work with genetics?

**Reply by TheRealQuantam** [reply]
They do according to Google. However, Pydantic doesn't support them (e.g. "Fields of type "<class 'module.GenericRange'>" are not supported." when you try to use them in a model or in parse_obj_as). Oddly, no error is thrown if you use a Pydantic dataclass with generics, but the fields that use the generic variables are not validated. E.g.
```
PosInt = conint(ge = 0, strict = True)
@dataclass
class GenericRange(Generic[T]):
	start: T
	end: T
goat = GenericRange[PosInt](-1, 2)
goat = parse_obj_as(GenericRange[PosInt], [-1, 2])
```

Both lines succeed and instantiate the object. The latter should fail validation because -1 < 0.

**Reply by samuelcolvin** [reply]
Is there an authoritative source saying they work? Eg the python docs? Otherwise do they work with mypy and/or pyright? 

"Google" doesn't mean much.

Short answer is we should be able to get them to work in v2, if they're a think. 

Please create a feature request, and include some proof that they work outside pydantic - eg they don't break at runtime and are supported by static type checkers.
